README.md
# Template-Plugin-Gettext

Localization for the Template Toolkit 2

## Description

This Perl library offers a complete translation solution for the Template Toolkit 2.
It consists of a plugin that offers translation functions inside templates
and a string extractor `xgettext-tt2` that extracts translatable strings
from templates and writes them to PO files (or rather a `.pot` file in PO
format).

## Usage

The solution offered by this library is suitable for templates that have 
a lot of markup (normally HTML) compared to text.  If the files contain
a lot of content other solutions are probably more suitable.  One of them
is [xml2po](https://github.com/mate-desktop/mate-doc-utils/tree/master/xml2po),
especially if the input format is HTML.

If the input format is Markdown, for example for a static side generator,
a feasible approach may be to simply split the input into paragraphs, and
turn each paragraph into an entry of a PO file.

In the following, we will assume that you have decided to localize
templates with this library.

### Templates

The first step is to mark all translatable strings.  This serves
a double purpose.  Strings are marked, so that the extractor 
`xgettext-tt2` can find them and write them into a translation file 
in PO format.

The second purpose is that these markers are also valid functions
resp. filters for the template toolkit and will interpolate the
translations for these messages into the output, when rendering the
template.  As a result, your templates remain pretty readable after
localizing them.

In every source file that you want to use translations, you have
to `USE` the template:

    [% USE gtx = Gettext('com.mydomain.www', 'fr') %]

Do *not* forget to `USE` the plug-in in all templates!  The template
toolkit will not warn you, when you forget it but the translation 
mechanism will not work!

The first argument is the so-called *textdomain*.  This is the
identifier for your message catalogs and also the basename of several
files.  In the example above, the translated message catalog would
be searched as *`LOCALEDIR`*`/fr/LC_MESSAGES/com.mydomain.www.mo`. The second parameter is the language.  This will normally come from
a variable instead of a hard-coded string.

A possible third argument (omitted in the example) is the character
set to use, all following arguments are additional directories to
search first for translations.

The default list of directories is:

* `./locale`
* `/usr/share/locale`
* `/usr/local/share/locale`

The directory `./locale` is relative to the current working directory
from where you invoke the template processor.

#### Simple Translations With `gettext()`

The simplest and most common way of doing things is:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

<title>[% gtx.gettext("World Of Themes") %]</title>
    
<h1>[% "Introduction" | gettext %]

<p>
[% FILTER gettext %]
The "World Of Themes" is the ultimate source of templates
for the Template Toolkit.
[% END %]
</p>
```

This shows three different ways of localizing strings.  You can
use the function `gtx.gettext()`, the filter `gettext` with pipe
syntax, or the same filter with block syntax.  The result is always
the same.  The string will be recognized as translatable by 
`xgettext-tt2` and it will be translated into the selected language,
when rendering the template.

#### Interpolating Strings Into Translations

One important thing to understand is that the argument to the
gettext functions or filters is the lookup key into the translation
database, when the template gets rendered.  That implies that this
key has to be invariable and must not use any interpolated variables.

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

[% gtx.gettext("Hello, $firstname $lastname!") %]
```

This template code is syntactically correct and will also render
correctly.  But `xgettext-tt2` will bail out on it with an error
message like

    templates.html:3: Illegal variable interpolation at "$"

The function `gettext()` will receive the interpolated string
as its argument, and that is not the same as the string that
the extractor program `xgettext-tt2` sees.  And that means that
the translation cannot be found.

The correct way to interpolate strings uses `xgettext()`:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

[% gtx.xgettext("Hello, {first} {last}!",
                first => firstname, last => lastname) %]
[% "Hello, {first} {last}!" | xgettext(first => firstname, 
                                       last => lastname) %]
[% FILTER xgettext(first => firstname, last => lastname) %]
Hello, {first} {last}!
[% END %]
```

One additional benefit of this is that the extractor program
`xgettext-tt2` will also mark these strings with the flag
"perl-brace-format".  When the translation from the `.po`
file gets compiled into an `.mo` file, the compiler `msgfmt`
checks that the translated strings contains exactly the same
placeholders as the original.

One thing that you should also avoid is to assemble strings
in the template source code.  Do *not*:

```html
[% gtx.gettext("Please contact") %] [% name %]
[% gtx.gettext("for help about the") %] [% package %]
[% gtx.gettext("software.") %]
```

This will result in three translatable text snippets
"Please contact", "for help about the", and "software." that
are hard to translate without context.  Besides it makes
illegal assumptions about the word order in translated sentences.
Instead, use `xgettext()` and write in complete sentences with
placeholders.

By the way, the `x` in the function `xgettext()` stands for *eXpand*
while the `x` in the program `xgettext-tt2` or GNU Gettext's
`xgettext` program stands for *eXtract*.

#### Plural Forms

Do *not* write this:

```html
[% IF num != 1 %]
[% gtx.xgettext("{number} documents deleted!", number => num) %]
[% ELSE %]
[% gtx.gettext("One document deleted!") %]
[% END %]
```

This assumes that every language has one singular and one plural
(and no other forms) and that the condition that selects the correct
form is always `COUNT != 1`.  But this is wrong for many languages
for example Russian (two plural forms), Chinese (no plural), French
(different condition), and many more.

Write instead:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

[% gtx.nxgettext("One document deleted.", 
                 "{count} documents deleted."
                 num,
                 count => num) %]
```

The function `nxgettext()` receives the singular and plural
form as the first and second argument, followed by the number
of items, followed by an arbitrary number of key/value pairs
for interpolating variables in the strings.

There is also a function `ngettext()` that does not expand
its first two arguments.  You will find out that you almost
never need that function.

You can also use `nxgettext()` and `ngettext()` as filters.
But the necessary code is awkward, and their use is therefore
not recommended.

#### Ambiguous Strings (message contexts)

Sometimes an English string has different meanings in other
languages:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

[% gtx.gettext("State:") %]
[% IF state == '1' %]
[% gtx.pgettext("state", "Open") %]
[% ELSE %]
[% gtx.gettext("Closed") %]
[% END %]
<a href="/action/open">[% gtx.pgettext("action", "Open") %]</a>
```

The function `pgettext()` works like gettext but has one 
extra argument preceding the string, the so-called
message context.  The string extractor `xgettext-tt2` will now
create two distinct messages "Open", one with the context "state",
the other one with the context "action".  The sole purpose of this
context is to disambiguate the string "Open" for languages where the
verb ("to open") and the adjective ("the door is *open*") has
two distinct translations.

You will normally use this function, when a translator asks you
to do so, but not on your own behalf.

There is also a function `pxgettext()` that supports placeholder
interpolation, and `npxgettext()` that has the following semantics:

```perl
npxgettext(CONTEXT, SINGULAR, PLURAL, COUNT,
           KEY1 => VALUE1, KEY2 => VALUE2, ...)
```

#### More Esoteric Functions

The [API documentation](lib/Template/Plugin/Gettext.pod#user-content-FUNCTIONS) contains
some more functions and filters that are available for completeness.
You will never need them in normal projects.

#### Translator Hints

You can add comments to the source code that are copied into the
`.po` file as hints for the translators.  This will look like
this:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

<!-- TRANSLATORS: This is the day of the week! -->
[% gtx.gettext("Sun") %]
```

In order to make that work, you have to invoke the extractor
program `xgettext-tt2` like this:

    xgettext-tt2 --add-comments=TRANSLATORS: t1.html t2.html ...

#### Modifying Flags

In rare situations, you may need the following:

```html
[% USE gtx = Gettext('com.mydomain.www', lang) %]

<!-- xgettext:no-perl-brace-format -->
[% gtx.xgettext("Value: {value}", value => whatever) %]
```

Normally, the argument of `xgettext()` will be flagged in
the `.po` file with "perl-brace-format", and a translation
will fail to compile if the translation does not contain exactly
the same placeholders as the original does.

You can override that default behavior for individual messages
by placing a comment containing the string "xgettext:" directly
in front of the string.

### Translation Workflow

The translation workflow is the standard workflow known from GNU 
Gettext.  All files relevant for translations are conventionally
kept in a subdirectory `po`.

You can save time if you use the seed project
[Template-Plugin-Gettext-Seed](https://github.com/gflohr/Template-Plugin-Gettext-Seed)
as a base.  It contains a directory `po` ready for use,
with --- at your choice --- a Makefile or a script `po-make.pl`
that automates the entire translation workflow.  It is also
prepared for extracting strings from other sources than
template files.  In that example, these are Perl source files,
but it will work in a similar fashion for other programming
languages.

But rolling your own version is also simple.  Just read on.

#### Extracting Strings With `xgettext-tt2`

Extracting translatable strings from templates for the Template
Toolkit 2 is as easy as:

```shell
$ xgettext-tt2 TEMPLATE....
```

This will scan all files given as arguments for translatable strings
and create a file `messages.po` with the strings found.

The normal invocation of `xgettext-tt2` is normally a little bit more
sophisticated:

```shell
$ xgettext-tt2 --files-from=POTFILES \
    --output=com.mydomain.www.pot \
    --add-comments=TRANSLATORS: --from-code=utf-8 \
    --force-po
```

You can, of course, write everyting in one line and omit the backslashes.

Specifying all input files as arguments on the command-line can
quickly become unwieldy.  It is more common to put the list of input
files into a text file, each input file on one line, and instruct
`xgettext-tt2` to read it with the option `--files-from`.  The name
of the file is by convention `POTFILES`.

The output file is normally a file `TEXTDOMAIN.pot`, where 
`TEXTDOMAION` is the identifier selected in the templates.  The
reverse hostname of the server serving the rendered templates
is a good choice.

If you want to be able to give hints to translators in the source
files, you have to specify the trigger string --- normally
"TRANSLATORS:" --- with the option `--add-comments`.  Specifying
an empty string (`--add-comments=''`) instructs `xgettext-tt2` 
to copy all comments into the `.pot` file.

If your templates contain characters outside of US-ASCII, you should
specify the character set of the template files with the option
`--from-code=CODESET`.

The option `--force-po` instructs `xgettext-tt2` to write an output
file even if no translatable strings had been found.  But this
is a matter of taste.  Omit the option, if you prefer it.

`xgettext-tt2` has a lot more options.  They are mostly compatible
with the ones of `xgettext` from GNU gettext for C, Perl, and
a lot more languages.  See the [documentation for GNU Gettext's
xgettext](https://www.gnu.org/software/gettext/manual/html_node/xgettext-Invocation.html)
and the documentation for [Locale::XGettext](https://github.com/gflohr/Locale-XGettext/blob/master/lib/Locale/XGettext.pod)
for more information.

By the way, why is the ouput file a `.pot` file and not a `.po`
file?  It is the *template* for the `.po` files for the individual
languages.  You never edit that file, but re-generate it, whenever
the source files have changed.  Hence, it only contains strings
in the original, in the base language.

#### Creating Translation Files

For each supported language (except for the base language) you
should create a file `LL.po`, where `LL` is the two-letter
language code for that language, for example `fr.po`, `de.po`,
or `it.po`.  You can also specify the combination of language
and country like in `de_DE.po` or `pt_BR.po`.

One option for that is to simply copy the `.pot` file and
edit the header accordingly.  It is normally easier to do that
with the program `msginit`:

```
$ msginit --input=com.mydomain.www.pot --locale=fr
```

Replace `TEXTDOMAIN.pot` with the name of the `.pot` file, and
`fr` with the language in question.  This will prefill a lot
of fields in the `.po` file.

#### Compiling Translation Files

The translated `.po` files are compiled with the program `msgfmt`:

```shell
$ msgfmt --check --statistics --verbose -o fr.mo fr.po
fr.po: 212 translated messages, 1 fuzzy translation, 3 untranslated messages.
```

This will compile the translation file `fr.po` into a binary
file `fr.mo`.  It also checks the translations for formal errors
and print statistics about the number of translated and
untranslated strings.

#### Installing Translation Files

The plugin does not use `.po` files for looking up translations
but the binary `.mo` files.  But it has to find them.

You have to decide for one of the directories that 
`Template::Plugin::Gettext` searches for translations.  The
default order is:

* `@INC/LocaleData`
* `/usr/share/locale`
* `/usr/local/share/locale`

The first line means that every directory `LocaleDir` inside
Perl's include directories is searched for translation files.
Keep in mind, that for security reasons the current directory
(`.`) is nowadays often *not* in Perl's `@INC`.

Let's assume that `/var/www/lib` is in Perl's @INC.  You would
then install the French translation file `fr.mo` as `/var/www/lib/LocaleData/fr/LC_MESSAGES/com.mydomain.www.mo`.
`TEXTDOMAIN` is a placeholder for the textdomain you have
selected (and `LC_MESSAGES` is *not* a placeholder but a real 
directory name).

That is good except for the fact that `/var/www/lib` is usually
not in Perl's `@INC`.  But you can change that where you invoke
the template processor:

```perl
BEGIN {
    unshift @INC, '/var/www/lib';
}

use Template;

Template->new->process('template.html', $data);
```

You can completely override the default search order in the
templates:

```html
[% USE gtx = Gettext('com.mydomain.www', lang, 'utf-8', 
                     '/var/www/locale', '/srv/www/locale')]
```

Now, the French translation would be searched in 
`/var/www/locale/fr/LC_MESSAGES/com.mydomain.www.mo` and 
`/var/www/locale/fr/LC_MESSAGES/com.mydomain.www.mo`.

#### Updating Translation Files

Translations may become obsolete, when the source templates
change.  In this case, you have to merge the new set of 
translatable strings into the existing translation files.
Fortunately, GNU Gettext makes this easy:

```shell
$ xgettext-tt2 --files-from=POTFILES \
    --output=com.mydomain.www.pot \
    --add-comments=TRANSLATORS: --from-code=utf-8 \
    --force-po
$ cp fr.po fr.old.po
$ msgmerge fr.old.po com.mydomain.www.pot -o fr.po
....... done
```

You first update the `.pot` file with `xgettext-tt2` so that it
contains the current set of translatable strings.  You then
make a backup of each `.po` file and then invoke the program
`msgmerge` for merging the current translations from `fr.old.po`
with the new set of strings from `com.mydomain.www.pot` into
the updated translation file `fr.po`.

The file `fr.po` will now contain the new strings as untranslated
entries.  Strings that have only slightly change will retain their
translations but they will be marked as "fuzzy", so that they
can be reviewed by a translator.  Entries for strings that are
no longer present in the sources are obsoleted.

#### Integrating With Other Programming Languages

The GNU Gettext framework is available for a lot of programming
languages and it is not uncommon that two or more of these 
languages are mixed in a project.  It is beneficial in these
cases to use a common translation base for all used 
technologies.

`xgettext-tt2` is based on [`Locale-XGettext`](https://github.com/gflohr/Locale-XGettext)
and therefore not only understands Template Toolkit templates
but also `.po` and `.pot` files as input.  GNU Gettext's xgettext
has the same feature.

Accumulating all translatable strings from the different
technologies is therefore very easy.  If you have a project
that uses Template Toolkit for rendering web pages and Perl
for the business logic you first extract strings from your 
Perl files --- as usual --- with `xgettext` from GNU gettext
into a temporary file, for example `plfiles.pot`.  Then you
extract the strings from the templates with `xgettext-tt2` 
from this library, but you specify `plfiles.pot` as an 
additional input file. And now the output file of `xgettext-tt2`
contains all the strings from the template files *plus* those
from the Perl files in `plfiles.pot`.

Of course, you can also do it the other way round, extract
with `xgettext-tt2` into `ttfiles.pot`, and then feed that as
an additional input file to GNU Gettext's `xgettext`.

You can use the seed project [Template-Plugin-Gettext-Seed](https://github.com/gflohr/Template-Plugin-Gettext-Seed)
as a fully functional starting point for such setups.

## Bugs

Please report bugs at [https://github.com/gflohr/Template-Plugin-Gettext/issues](https://github.com/gflohr/Template-Plugin-Gettext/issues)

## Author

Template-Plugin-Gettext was written by [Guido Flohr](http://www.guido-flohr.net/).


Changes
{{$NEXT}}
	- The module is now considered stable. Version bumped to 1.0.
	- Setting the environment variable $LANGUAGE no longer prevents
	  the module from working correctly (github issue
	  https://github.com/gflohr/Template-Plugin-Gettext/issues/18)
	- Other ssues fixed:
	  o https://github.com/gflohr/Template-Plugin-Gettext/issues/11
	  o https://github.com/gflohr/Template-Plugin-Gettext/issues/13
	  o https://github.com/gflohr/Template-Plugin-Gettext/issues/14

0.8  2023-03-15
	- Fixed spelling errors.
	- Documentation.

0.7  2018-11-04
        - Always set utf8 flag on translations.
        - Require libintl-perl 1.30, because of
          https://github.com/gflohr/libintl-perl/issues/4

0.6  2018-05-09
        - Re-allow relative paths (was broken in 0.5).  Otherwise common
          POTFILES files cannot be read.
        - Allow argument to --plug-in to be empty so that you can extract
          arguments from functions in the TT2 root namespace. 

0.5  2018-03-27
        - Template::Plugin::Gettext now automatically sets the utf8 flag
          on translated strings, when the original string (the msgid) had
          the utf8 flag set.
        - xgettext-tt2 can now be used without a plug-in, i. e. for functions
          in the root namespace.
        - Issues fixed:
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/6
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/7
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/8
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/9

0.4  2018-01-23
        - Issues fixed:
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/2
          o https://github.com/gflohr/Template-Plugin-Gettext/issues/4

0.3  2018-01-02
        - New option '--plug-in' allows overriding the plug-in name,
          see (https://github.com/gflohr/Template-Plugin-Gettext/issues/1)

0.2  2017-12-31
        - Extract nested method invocations.
        - Remember all files using a certain textdomain.

0.1 2017-09-10
        - Initial version.


t/TestLib.pm
#! /usr/bin/env perl

# Copyright (C) 2016 Guido Flohr <guido.flohr@cantanea.com>,
# all rights reserved.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published
# by the Free Software Foundation; either version 2, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.

# You should have received a copy of the GNU Library General Public
# License along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
# USA.

use strict;

use vars qw(@ISA @EXPORT_OK);

use Locale::XGettext::TT2;
use Locale::XGettext::Util::Keyword;

@ISA = qw(Exporter);
@EXPORT_OK  = qw(find_entries use_keywords);

sub find_entries {
    my ($po, %args) = @_;

    my @hits;
    foreach my $entry (@$po) {
        next if exists $args{msgid} && $entry->msgid ne $args{msgid};
        next if exists $args{msgstr} && $entry->msgstr ne $args{msgstr};
        next if exists $args{msgctxt} && $entry->msgctxt ne $args{msgctxt};
        next if exists $args{comment} && $entry->comment ne $args{comment};
        next if exists $args{automatic} && $entry->automatic ne $args{automatic};
        push @hits, $entry;
    }

    return @hits;
}

sub use_keywords($) {
    my ($keywords) = @_;

    foreach my $method (keys %$keywords) {
        $keywords->{$method} =
            Locale::XGettext::Util::Keyword->new($method,
                                                 @{$keywords->{$method}});
    }

    Locale::XGettext::TT2->new({keywords => $keywords}, 'dummy');
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.030.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Gettext Support For the Template Toolkit Version 2",
  "dist_author" => [
    "Guido Flohr <guido.flohr\@cantanea.com>"
  ],
  "dist_name" => "Template-Plugin-Gettext",
  "dist_version" => "0.8",
  "license" => "lgpl",
  "module_name" => "Template::Plugin::Gettext",
  "recursive_test_files" => 1,
  "requires" => {
    "Cwd" => 0,
    "Encode" => 0,
    "Getopt::Long" => 0,
    "Locale::Messages" => 0,
    "Locale::TextDomain" => "1.30",
    "Locale::Util" => 0,
    "Locale::XGettext" => "0.7",
    "Template" => 0,
    "Template::Parser" => 0,
    "Template::Plugin" => 0,
    "base" => 0,
    "strict" => 0
  },
  "script_files" => [
    "bin/xgettext-tt2"
  ],
  "test_requires" => {
    "File::Spec" => 0,
    "Locale::XGettext::Util::Keyword" => 0,
    "Test::More" => 0,
    "vars" => 0
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "Locale::XGettext::Util::Keyword" => 0,
  "Module::Build" => "0.28",
  "Test::More" => 0,
  "vars" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.030.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Gettext Support For the Template Toolkit Version 2",
  "AUTHOR" => "Guido Flohr <guido.flohr\@cantanea.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Template-Plugin-Gettext",
  "EXE_FILES" => [
    "bin/xgettext-tt2"
  ],
  "LICENSE" => "lgpl",
  "NAME" => "Template::Plugin::Gettext",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "Encode" => 0,
    "Getopt::Long" => 0,
    "Locale::Messages" => 0,
    "Locale::TextDomain" => "1.30",
    "Locale::Util" => 0,
    "Locale::XGettext" => "0.7",
    "Template" => 0,
    "Template::Parser" => 0,
    "Template::Plugin" => 0,
    "base" => 0,
    "strict" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "Locale::XGettext::Util::Keyword" => 0,
    "Test::More" => 0,
    "vars" => 0
  },
  "VERSION" => "0.8",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Cwd" => 0,
  "Encode" => 0,
  "File::Spec" => 0,
  "Getopt::Long" => 0,
  "Locale::Messages" => 0,
  "Locale::TextDomain" => "1.30",
  "Locale::Util" => 0,
  "Locale::XGettext" => "0.7",
  "Locale::XGettext::Util::Keyword" => 0,
  "Module::Build" => "0.28",
  "Template" => 0,
  "Template::Parser" => 0,
  "Template::Plugin" => 0,
  "Test::More" => 0,
  "base" => 0,
  "strict" => 0,
  "vars" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Locale/XGettext/TT2/Script.pm
#! /bin/false

# Copyright (C) 2016 Guido Flohr <guido.flohr@cantanea.com>,
# all rights reserved.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published
# by the Free Software Foundation; either version 2, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.

# You should have received a copy of the GNU Library General Public
# License along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
# USA.

package Locale::XGettext::TT2::Script;

use strict;

use base 'Exporter';
use vars qw(@EXPORT $VERSION);
@EXPORT = qw($VERSION);
$VERSION = '0.1.1';

use Locale::TextDomain qw(Template-Plugin-Gettext);
use Getopt::Long qw(GetOptionsFromArray);

use Locale::XGettext::TT2;
use Locale::XGettext::TT2::Keyword;

use base qw(Locale::XGettext::TT2);

sub new {
    my ($class, $argv) = @_;

    my $self = {};
    bless $self, $class;
    my %options = $self->__getOptions($argv);
    
    $self->__displayUsage if $options{help};
    $self->__displayVersion if $options{version};
    
    $options{keyword} = $self->__setKeywords($options{keyword});

    return $class->SUPER::new(\%options, @$argv);
}

sub run {
	my ($self) = @_;
	
	eval { $self->SUPER::run() };
	if ($@) {
		chomp $@;
	    die "$0: $@\n";
	}
	
	return $self;
}

sub output {
	my ($self) = @_;

    eval { $self->SUPER::output() };
    if ($@) {
        chomp $@;
        die "$0: $@\n";
    }
    
    return $self;
}

sub __getOptions {
    my ($self, $argv) = @_;
    
    my %options;
    
    Getopt::Long::Configure('bundling');
    GetOptionsFromArray($argv,
        # Input file location:
        'f|files-from=s@' => \$options{files_from},
        'D|directory=s@' => \$options{directory},

        # Output file location:
        'd|default-domain=s' => \$options{default_domain},
        'o|output=s' => \$options{output},
        'p|output-dir=s' => \$options{output_dir},

        # Input file interpretation.
        'from-code=s' => \$options{from_code},
       
        # Operation mode:
        'j|join-existing' => \$options{join_existing},
        # FIXME! Is this allowed multiple times?
        'x|exclude-file=s' => \$options{exclude_file},
        'c|add-comments:s@' => \$options{add_comments},
        'check=s@' => \$options{check},
        'sentence_end=s' => \$options{sentence_end},

        # Language specific options:
        'a|extract-all' => \$options{extract_all},
        'k|keyword:s@' => \$options{keyword},
         
        # Output details:
        'force-po' => \$options{force_po},
        'no-location' => \$options{no_location},
        'n|add-location' => \$options{add_location},
        's|sort-output' => \$options{sort_output},
        'F|sort-by-file' => \$options{sort_by_file},
        'omit-header' => \$options{omit_header},
        'copyright-holder=s' => \$options{copyright_holder},
        'foreign-user' => \$options{foreign_user},
        'package_name=s' => \$options{package_name},
        'package_version=s' => \$options{package_version},
        'msgid-bugs-address=s' => \$options{msgid_bugs_address},
        'm|msgstr-prefix:s' => \$options{msgid_str_prefix},
        'M|msgstr-suffix:s' => \$options{msgid_str_suffix},

        # Informative output.
        'h|help' => \$options{help},
        'V|version' => \$options{version},
    ) or exit 1;
    
    foreach my $key (keys %options) {
    	delete $options{$key} if !defined $options{$key};
    }
    
    return %options;   
}

sub __setKeywords {
    my ($self, $options) = @_;
    
    my %keywords = Locale::XGettext::TT2->defaultKeywords;
    while (my ($method, $argspec) = each %keywords) {
    	$keywords{$method} = Locale::XGettext::TT2::Keyword->new($method, @$argspec);
    }
    
    foreach my $option (@$options) {
    	if ('' eq $option) {
    	    undef %keywords;
    	    next;
    	}

        my $keyword = Locale::XGettext::TT2::Keyword->newFromString($option);
        $keywords{$keyword->method} = $keyword;
    }

    return \%keywords;
}

sub __displayUsage {
    print __x("Usage: {program} [OPTION] [INPUTFILE]...\n", program => $0);
    
    print "\n";
    
    print __(<<EOF);
Extract translatable strings from given input files.  
EOF
    print __(<<EOF);
Input files are interpreted as Template Toolkit version 2 templates.  
EOF

    print "\n";
    
    print __(<<EOF);
Mandatory arguments to long options are mandatory for short options too.
Similarly for optional arguments.
EOF

    print "\n";

    print __(<<EOF);
Input file location:
EOF

    print __(<<EOF);
  INPUTFILE ...               input files
EOF

    print __(<<EOF);
  -f, --files-from=FILE       get list of input files from FILE\
EOF

    print __(<<EOF);
  -D, --directory=DIRECTORY   add DIRECTORY to list for input files search
EOF

    printf __(<<EOF);
If input file is -, standard input is read.
EOF

    print "\n";

    printf __(<<EOF);
Output file location:
EOF

    printf __(<<EOF);
  -d, --default-domain=NAME   use NAME.po for output (instead of messages.po)
EOF

    print __(<<EOF);
  -o, --output=FILE           write output to specified file
EOF

    print __(<<EOF);
  -p, --output-dir=DIR        output files will be placed in directory DIR
EOF

    print __(<<EOF);
If output file is -, output is written to standard output.
EOF

    print "\n";

    print __(<<EOF);
Input file interpretation:
EOF

    print __(<<EOF);
      --from-code=NAME        encoding of input files
EOF
    print __(<<EOF);
By default the input files are assumed to be in ASCII.
EOF

    printf "\n";

    print __(<<EOF);
Operation mode:
EOF

    print __(<<EOF);
  -j, --join-existing         join messages with existing file
EOF

    print __(<<EOF);
  -x, --exclude-file=FILE.po  entries from FILE.po are not extracted
EOF

    print __(<<EOF);
  -cTAG, --add-comments=TAG   place comment blocks starting with TAG and
                                preceding keyword lines in output file
  -c, --add-comments          place all comment blocks preceding keyword lines
                                in output file
EOF

    print __(<<EOF);
      --check=NAME            perform syntax check on messages
                                (ellipsis-unicode, space-ellipsis,
                                 quote-unicode, bullet-unicode)
EOF
    print __(<<EOF);
      --sentence-end=TYPE     type describing the end of sentence
                                (single-space, which is the default,
                                 or double-space)
EOF

    print "\n";

    print __(<<EOF);
Language specific options:
EOF

    print __(<<EOF);
  -a, --extract-all           extract all strings
EOF

    print __(<<EOF);
  -kWORD, --keyword=WORD      look for WORD as an additional keyword
  -k, --keyword               do not to use default keywords"));
EOF

    # FIXME! Other plug-in name? Or combine it with the keywords, like
    # --keyword=Maketext.blabla?

    print "\n";

    print __(<<EOF);
Output details:
EOF

    print __(<<EOF);
      --force-po              write PO file even if empty
EOF

    print __(<<EOF);
      --no-location           do not write '#: filename:line' lines
EOF

    print __(<<EOF);
  -n, --add-location          generate '#: filename:line' lines (default)
EOF

    print __(<<EOF);
  -s, --sort-output           generate sorted output
EOF

    print __(<<EOF);
  -F, --sort-by-file          sort output by file location
EOF

    print __(<<EOF);
      --omit-header           don't write header with 'msgid ""' entry
EOF

    print __(<<EOF);
      --copyright-holder=STRING  set copyright holder in output
EOF

    print __(<<EOF);
      --foreign-user          omit FSF copyright in output for foreign user
EOF

    print __(<<EOF);
      --package-name=PACKAGE  set package name in output
EOF

    print __(<<EOF);
      --package-version=VERSION  set package version in output
EOF

    print __(<<EOF);
      --msgid-bugs-address=EMAIL\@ADDRESS  set report address for msgid bugs
EOF

    print __(<<EOF);
  -m[STRING], --msgstr-prefix[=STRING]  use STRING or "" as prefix for msgstr
                                values
EOF

    print __(<<EOF);
  -M[STRING], --msgstr-suffix[=STRING]  use STRING or "" as suffix for msgstr
                                values
EOF

    printf "\n";

    print __(<<EOF);
Informative output:
EOF

    print __(<<EOF);
  -h, --help                  display this help and exit
EOF

    print __(<<EOF);
  -V, --version               output version information and exit
EOF

    printf "\n";

    # TRANSLATORS: The placeholder indicates the bug-reporting address
    # for this package.  Please add _another line_ saying
    # "Report translation bugs to <...>\n" with the address for translation
    # bugs (typically your translation team's web or email address).
    print __"Report bugs at <https://github.com/gflohr/template-plugin-gettext/issues>!\n";

    exit 0;
}

sub __usageError {
    my $message = shift;
    if ($message) {
        $message =~ s/\s+$//;
        $message = "$0: $message\n";
    }
    else {
        $message = '';
    }
    die <<EOF;
${message}Usage: $0 [OPTIONS]
Try '$0 --help' for more information!
EOF
}

sub __displayVersion {
    my $msg = __x('{program} (Template-Plugin-Gettext) {version}
Copyright (C) {years} Cantanea EOOD (http://www.cantanea.com/).
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Written by Guido Flohr (http://www.guido-flohr.net/).
', program => $0, years => 2016, version => $Locale::XGettext::TT2::VERSION);

    print $msg;

    exit 0;
}

1;
